File: python-pr/52/python/code-quality-readability-maintainability/file_organizer_41.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
def categorize_files_by_extension(directory):
    import os
    
    file_categories = {
        'images': ['jpg', 'jpeg', 'png', 'gif', 'bmp'],
        'documents': ['pdf', 'doc', 'docx', 'txt', 'rtf'],
        'videos': ['mp4', 'avi', 'mov', 'wmv', 'flv']
    }
    
    categorized_files = {category: [] for category in file_categories}
    categorized_files['others'] = []
    
    files = os.listdir(directory)
    
    for filename in files:
        file_extension = filename.split('.')[-1].lower()
        category = _get_file_category(file_extension, file_categories)
        categorized_files[category].append(filename)
    
    return categorized_files

def _get_file_category(extension, categories):
    for category, extensions in categories.items():
        if extension in extensions:
            return category
    return 'others'
# {/fact}

File: js-pr/56/javascript/code-quality-readability-maintainability/file_organizer_41.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
function categorizeFilesByExtension(directory) {
    const fs = require('fs');
    
    const fileCategories = {
        images: ['jpg', 'jpeg', 'png', 'gif', 'bmp'],
        documents: ['pdf', 'doc', 'docx', 'txt', 'rtf'],
        videos: ['mp4', 'avi', 'mov', 'wmv', 'flv']
    };
    
    const categorizedFiles = {
        images: [],
        documents: [],
        videos: [],
        others: []
    };
    
    try {
        const files = fs.readdirSync(directory);
        
        files.forEach(filename => {
            const extension = getFileExtension(filename);
            const category = determineFileCategory(extension, fileCategories);
            categorizedFiles[category].push(filename);
        });
        
        return categorizedFiles;
    } catch (error) {
        console.error('Error organizing files:', error.message);
        return categorizedFiles;
    }
}

function getFileExtension(filename) {
    return filename.split('.').pop().toLowerCase();
}

function determineFileCategory(extension, categories) {
    for (const [category, extensions] of Object.entries(categories)) {
        if (extensions.includes(extension)) {
            return category;
        }
    }
    return 'others';
}
// {/fact}